src/topology/uniform_space/basic.lean

```diff
@@ -599,6 +699,52 @@ end
 ### Closure and interior in uniform spaces
 -/
 
+lemma closure_eq_uniformity (s : set $ α × α) :
+  closure s = ⋂ V ∈ {V | V ∈ 𝓤 α ∧ symmetric_rel V}, V ○ s ○ V :=
+begin
+  ext ⟨x, y⟩,
+  simp_rw [mem_closure_iff_nhds_basis (uniform_space.has_basis_nhds_prod x y),
+           mem_Inter, mem_set_of_eq],
+  apply forall_congr,
+  intro V,
+  apply forall_congr,
+  rintros ⟨V_in, V_symm⟩,
+  simp_rw [mem_comp_comp V_symm, inter_comm, exists_prop],
+  exact iff.rfl,
+end
+
+lemma uniformity_has_basis_closed : has_basis (𝓤 α) (λ V : set (α × α), V ∈ 𝓤 α ∧ is_closed V) id :=
+begin
+  rw filter.has_basis_self,
+  intro t,
+  split,
+  { intro h,
+    rcases comp_comp_symm_mem_uniformity_sets h with ⟨w, w_in, w_symm, r⟩,
+    refine ⟨closure w, _,  is_closed_closure, _⟩,
+    apply mem_sets_of_superset w_in subset_closure,
+    refine subset.trans _ r,
+    rw closure_eq_uniformity,
+    apply Inter_subset_of_subset,
+    apply Inter_subset,
+    exact ⟨w_in, w_symm⟩ },
+  { rintros ⟨r, r_in, r_closed, r_sub⟩,
+    exact mem_sets_of_superset r_in r_sub, }
+end
+
+lemma uniformity_has_basis_closure : has_basis (𝓤 α) (λ V : set (α × α), V ∈ 𝓤 α) closure :=
```

Comment:
A docstring on this one would be welcome, as the `has_basis ... closure` is surprising at first, before one tries to remember what `has_basis` really says.